drivers/vdpa/vfe/vfe_vdpa.c

```diff
@@ -0,0 +1,1196 @@
+ * Copyright 2019 Mellanox Technologies, Ltd
+ */
+#include <unistd.h>
+#include <net/if.h>
+#include <sys/socket.h>
+#include <sys/ioctl.h>
+#include <fcntl.h>
+#include <netinet/in.h>
+
+#include <rte_malloc.h>
+#include <rte_log.h>
+#include <rte_errno.h>
+#include <rte_string_fns.h>
+#include <rte_bus_pci.h>
+#include <rte_vfio.h>
+#include "rte_vhost.h"
+#include <rte_vdpa.h>
+#include <vdpa_driver.h>
+
+#include <rte_kvargs.h>
+
+#include "virtio_pci.h"
+#include "virtqueue.h"
+
+RTE_LOG_REGISTER(vfe_vdpa_logtype, pmd.vdpa.vfe, NOTICE);
+#define DRV_LOG(level, fmt, args...) \
+	rte_log(RTE_LOG_ ## level, vfe_vdpa_logtype, \
+		"VFE %s(): " fmt "\n", __func__, ##args)
+
+struct virtio_hw_internal vfe_virtio_hw_internal[VFE_MAX_PORT_NUM];
+
+enum {
+	VFE_VDPA_NOTIFIER_STATE_DISABLED,
+	VFE_VDPA_NOTIFIER_STATE_ENABLED,
+	VFE_VDPA_NOTIFIER_STATE_ERR
+};
+
+struct vfe_vring_info {
+	uint64_t desc;
+	uint64_t avail;
+	uint64_t used;
+	uint16_t size;
+	uint16_t last_avail_idx;
+	uint16_t last_used_idx; /*other is for future use, not use now*/
+	bool enable;
+	uint8_t notifier_state;
+	uint16_t index;
+	struct rte_intr_handle *intr_handle;
+	struct vfe_vdpa_priv *priv;
+};
+
+#define VFE_VDPA_DRIVER_NAME vdpa_vfe
+struct vfe_vdpa_priv {
+	TAILQ_ENTRY(vfe_vdpa_priv) next;
+	struct rte_pci_device *pdev;
+	struct rte_vdpa_device *vdev;
+	struct virtio_pci_dev vpdev;
+	int vfio_container_fd;
+	int vfio_group_fd;
+	int vfio_dev_fd;
+	int vid;
+	uint64_t guest_features;
+	pthread_mutex_t vq_config_lock;
+//	  virtio_admin_pf *manger_pf;  // for admin pf op
+	int configured;
+	uint16_t nr_virtqs;   //number of vq vhost enabled 
+	uint16_t hw_nr_virtqs; // number of vq hardware supported
+	struct vfe_vring_info **vrings;
+
+};
+
+//TO_DO: need to addd VHOST_USER_PROTOCOL_F_CONFIG later
+#define VFE_VDPA_PROTOCOL_FEATURES \
+				((1ULL << VHOST_USER_PROTOCOL_F_SLAVE_REQ) | \
+				 (1ULL << VHOST_USER_PROTOCOL_F_SLAVE_SEND_FD) | \
+				 (1ULL << VHOST_USER_PROTOCOL_F_HOST_NOTIFIER) | \
+				 (1ULL << VHOST_USER_PROTOCOL_F_REPLY_ACK))
+/*				   (1ULL << VHOST_USER_PROTOCOL_F_LOG_SHMFD) | \
+				 (1ULL << VHOST_USER_PROTOCOL_F_MQ) | \
+				 (1ULL << VHOST_USER_PROTOCOL_F_NET_MTU) | \
+				 (1ULL << VHOST_USER_PROTOCOL_F_STATUS))
+*/
+
+TAILQ_HEAD(vfe_vdpa_privs, vfe_vdpa_priv) vfe_priv_list =
+						  TAILQ_HEAD_INITIALIZER(vfe_priv_list);
+static pthread_mutex_t priv_list_lock = PTHREAD_MUTEX_INITIALIZER;
+
+static struct vfe_vdpa_priv *
+vfe_vdpa_find_priv_resource_by_vdev(struct rte_vdpa_device *vdev)
+{
+	struct vfe_vdpa_priv *priv;
+	int found = 0;
+
+	pthread_mutex_lock(&priv_list_lock);
+	TAILQ_FOREACH(priv, &vfe_priv_list, next) {
+		if (vdev == priv->vdev) {
+			found = 1;
+			break;
+		}
+	}
+	pthread_mutex_unlock(&priv_list_lock);
+	if (!found) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s.", vdev->device->name);
+		rte_errno = EINVAL;
+		return NULL;
+	}
+	return priv;
+}
+
+static uint16_t
+vfe_get_nr_vq(struct virtio_hw *hw)
```

Comment:
to avoid mix of api conventions,

we want to do

module_submodule_object_verb() notion.

for example,
virtio_vdpa_vqs_max_get()
virtio_vdpa = module
vqs = objects
max_get = verb